rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,737 @@
+#### Update
+
+The `Update` RPC takes a resource to be updated and must also return the updated resource so that any fields that are
+populated server side are provided to clients without requiring an additional call to `Get`. If partial updates of a
+resource are desired, the request may contain
+a [FieldMask](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#field-mask).
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Updates an existing Foo in the backend.
+    rpc UpdateFoo(UpdateFooRequest) returns (UpdateFooResponse);
+
+message UpdateFooRequest {
+  // The full Foo resource to update in the backend.
+  Foo foo = 1;
+  // A partial update for an existing Foo resource.
+  FieldMask update_mask = 2;
+}
+
+message UpdateFooResponse {
+  // The updated Foo as it exists in the backend. The only differences between the returned
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+#### Upsert
+
+> The `Create` and `Update` RPCs should be preferred over `Upsert` for normal operations,
+> see [#1326](https://github.com/gravitational/teleport/issues/1326) for more details.
+
+The `Upsert` RPC takes a resource that will overwrite a matching existing resource or create a new resource if one does
+not exist. The upserted resource is returned so that any fields that are populated server side are provided to clients
+without requiring a call to `Get`. If `Upsert` is not consumed it may be omitted from the API in favor of `Create` and
+`Update`.
+
+```protobuf
+// Creates a new Foo or replaces an existing Foo in the backend.
+    rpc UpsertFoo(UpsertFooRequest) returns (UpsertFooResponse);
+
+message UpsertFooRequest {
+  // The full Foo resource to persist in the backend.
+  Foo foo = 1;
+}
+
+message UpsertFooResponse {
+  // The upserted Foo as it exists in the backend. The only differences between the returned
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+#### Get
+
+The `Get` RPC takes the parameters required to match a resource (usually the resource name should suffice), and returns
+the matched resource.
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Returns a single Foo matching the request
+    rpc GetFoo(GetFooRequest) returns (GetFooResponse);
+
+message GetFooRequest {
+  // A filter to match the Foo by. Some resource may require more parameters to match and
+  // may not use the name at all.
+  string foo_id = 1;
+}
+
+message GetFooResponse {
+  // The matched Foo resource.
+  Foo foo = 1;
+}
+```
+
+#### List
```

Comment:
I think we should add a note regarding the expected ordering.
Resource's `name` field ascending, perhaps.

Even if we add a disclaimer saying that clients must not depend on the ordering.

Reply:
If there's pagination then there must be some order. I think the order guarantees should be specific by each RPC, but it's a good idea to encourage APIs to make explicit decisions (as in, say whether there is a guaranteed order or not).